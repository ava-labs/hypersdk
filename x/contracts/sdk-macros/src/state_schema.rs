// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

use proc_macro2::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    spanned::Spanned,
    token, Attribute, Error, Fields, Ident, Token, Type, Visibility,
};

pub fn impl_state_schema(
    key_pairs: impl IntoIterator<Item = KeyPair>,
) -> Result<TokenStream, Error> {
    key_pairs.into_iter().enumerate().try_fold(
        quote! {},
        |mut token_stream,
         (
            i,
            KeyPair {
                key_comments,
                key_vis,
                key_type_name,
                key_fields,
                value_type,
            },
        )| {
            let i = u8::try_from(i).map_err(|_| {
                Error::new(
                    key_type_name.span(),
                    "Cannot exceed `u8::MAX + 1` keys in a state-schema",
                )
            })?;

            token_stream.extend(Some(quote! {
                #(#key_comments)*
                #[derive(Copy, Clone, wasmlanche::bytemuck::NoUninit)]
                #[bytemuck(crate = "wasmlanche::bytemuck")]
                #[repr(C, packed)]
                #key_vis struct #key_type_name #key_fields;

                wasmlanche::prefixed_key_size_check!(wasmlanche::macro_types, #key_type_name);

                unsafe impl wasmlanche::macro_types::Schema for #key_type_name {
                    type Value = #value_type;

                    fn prefix() -> u8 {
                        #i
                    }
                }
            }));

            Ok(token_stream)
        },
    )
}

#[derive(Debug)]
pub struct KeyPair {
    key_comments: Vec<Attribute>,
    key_vis: Visibility,
    key_type_name: Ident,
    key_fields: Fields,
    value_type: Type,
}

impl Parse for KeyPair {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let key_comments = input.call(Attribute::parse_outer)?;
        let key_vis = input.parse::<Visibility>()?;
        let key_type_name: Ident = input.parse()?;
        let lookahead = input.lookahead1();

        let key_fields = if lookahead.peek(token::Paren) {
            let fields = input.parse()?;
            Fields::Unnamed(fields)
        } else if lookahead.peek(token::Brace) {
            Fields::Named(input.parse()?)
        } else {
            Fields::Unit
        };

        if let Fields::Named(named) = key_fields {
            return Err(Error::new(
                named.span(),
                "types with named fields are not supported",
            ));
        }

        input.parse::<Token![=>]>()?;
        let value_type = input.parse()?;

        Ok(Self {
            key_comments,
            key_vis,
            key_type_name,
            key_fields,
            value_type,
        })
    }
}
