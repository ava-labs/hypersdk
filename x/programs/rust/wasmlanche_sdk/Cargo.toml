[package]
name = "wasmlanche_sdk"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
thiserror = "1.0.46"
sdk_macros = { version = "0.1.0", path = "../sdk_macros" }
borsh = { version = "1.2.0", features = ["derive"] }

# TODO: delete serde dependencies(simulator needs update) 
serde = { version = "1.0.185", features = ["derive"] }
serde_bare = "0.5.0"

# simulator
serde_json = { version = "1.0.107", optional = true }

[features]
default = []
simulator = ["serde_json"]

[lib]
