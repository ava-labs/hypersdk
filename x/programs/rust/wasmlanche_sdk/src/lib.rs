#![deny(clippy::pedantic)]

pub mod errors;
pub mod host;
pub mod store;
pub mod types;
pub use sdk_macros::public;
