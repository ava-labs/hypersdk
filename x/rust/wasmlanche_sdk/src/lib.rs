pub mod errors;
pub mod host;
pub mod program;
pub mod store;
pub mod types;
