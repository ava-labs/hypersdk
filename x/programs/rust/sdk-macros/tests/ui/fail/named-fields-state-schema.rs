use wasmlanche_sdk::state_schema;

state_schema! {
    Valid => u8,
    OtherValid(u8) => u8,
    Invalid { param: u8 } => u8,
}

fn main() {}
