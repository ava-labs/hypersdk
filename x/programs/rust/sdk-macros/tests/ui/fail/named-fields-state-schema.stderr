error: types with named fields are not supported
 --> tests/ui/fail/named-fields-state-schema.rs:6:13
  |
6 |     Invalid { param: u8 } => u8,
  |             ^^^^^^^^^^^^^
